### DISCUSSION TITLE
Separate directory for configuration and data

### DISCUSSION THREAD

**Discussion by octvs** [GothenburgBitFactory/timewarrior#578] [Q&A]
Just started using timewarrior, thanks for the effort.

I noticed having a separate environment variable such as `TIMEWARRIORRC` to 
separate the user config from the data files came up several times [1-2].
Although this is addressed somewhat with XDG Base Directory implementation [3], 
it doesn't let the user choose a different directory to back-up or sync
the data. Although it is trivial to do

```sh
ln -s $XDG_DATA_HOME/timewarrior/data <path-desired>
```

Wouldn't it be better to let user choose whichever place to store data (without
configuration files), via TIMEWARRIORDB (or alternatively with a configuration
option on current default `$XDG_CONFIG_HOME/timewarrior/timewarrior.cfg`) ? 

I yet to see a need to use config file, but wouldn't this allow people to share
the user configuration between instances and only check data to their
back-up/sync solution?

1: https://github.com/GothenburgBitFactory/timewarrior/issues/193#issue-387064790, 3rd point
2: https://github.com/GothenburgBitFactory/timewarrior/issues/392#issuecomment-709506452
3: https://github.com/GothenburgBitFactory/timewarrior/issues/207

**Comment by lauft** [comment]
I am not quite sure whether I understand your suggestion.

With #207 the user can now choose to store their `timewarrior.cfg` file under `XDG_CONFIG_HOME` and their data files separate under `XDG_DATA_HOME`. With that the user can backup their data independently from the configuration.

Are you proposing to introduce an environment variable `TIMEWARRIORRC` to specify the location of the configuration file?

**Reply by octvs** [reply]
Sorry for the late response, life got in the way and I just circumvented this
via consistent symlinks. However I still see merit on my proposition.

> Are you proposing to introduce an environment variable `TIMEWARRIORRC` to
> specify the location of the configuration file?

~No, not really. I was asking for letting arbitrary locations for data files to~
~be defined. For example it can be ~
~- an environment variable `TIMEWARRIORDB` _notice that the last two letters are~
~  not the same_. ~
~- or adding a command-line flag `-f <file>`~
~- or adding a configuration option `file=<file>` to the `timewarrior.cfg` file~
~  where user can define a db file~

~Why? I myself want to store the data somewhere else, as well as have two~
~totally separate databases. One I can share with others via git, the other can~
~be personal etc.~                                                        

If example/inspirations are welcome: ledger[1] is a program that stood the test
of time, let's users to do something similar[2].

[1]: https://ledger-cli.org/index.html
[2]: https://ledger-cli.org/doc/ledger3.html#index-_002df-FILE-2

Cheers

**Reply by smemsh** [reply]
there is already a `$TIMEWARRIORDB` and I am using it in my shell profile for a long time.  It's described in section 7 of the `timew-config` manual.

I probably would agree that having both a section 1 and 7 man page for this is confusing, and could effectively hide it, but nonetheless, it's there.

**Reply by octvs** [reply]
I've noticed even I forgot my initial point since my original comment was
months ago. Using `TIMEWARRIORDB` env var, I saw `timewarrior.cfg` show up on
my git workspace, circulling me back to where I was when I opened this issue.

I was completely aware of the environment variable initially, as obvious from
the initial post. The idea was to have an option to separate the location of
config file separate from the data files. I will strikethrough my comment above
to avoid futher confusion.

However, I need to point out that this is not really required, when I wrote
this I expected timewarrior to require some configuration, but all these time I
have been using it I never touched to the config file. So I will just append it
to gitignore. My apologies for the noise.

**Reply by smemsh** [reply]
There are indeed two separate variables for data and config, to get what you were asking originally.  `$XDG_DATA_HOME` and `$XDG_CONFIG_HOME`.  Mentioned in man page, and earlier comments here.  In any case, glad you are fully functional.

**Reply by octvs** [reply]
> There are indeed two separate variables for data and config, to get what you
> were asking originally.  `$XDG_DATA_HOME` and `$XDG_CONFIG_HOME`.  

That was never the question. It was always possible with

```sh
alias timew='XDG_DATA_HOME=<custom_dir> timew'
```

My post specifically says "wouldn't it be better to ...". Alas, thanks for your
time.

**Reply by smemsh** [reply]
I really don't understand, maybe I'm slow.  You asked:

> Wouldn't it be better to let user choose whichever place to store data (without
configuration files

This is already possible using aforementioned environment variables.   No configuration file needed.  What am I missing? It seems it was always the question from the beginning, rather than "never."

**Reply by octvs** [reply]
> No configuration file needed.

"without configuration files" implies a separation of storage between config
and data, _just as the title_. 

> This is already possible using aforementioned environment variables.

If it were not; it would have been on "Issues", this is an issue of convenience
therefore "Discussions".

> What am I missing?

something to do?

**Reply by smemsh** [reply]
Ok, so the whole thing was tautological.  Got it :heavy_check_mark:
